File: internal/adapters/sources/hatena_fetcher.go
package sources

import (
	"context"
	"errors"
	"io"
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/HiroVodka/trendog/internal/domain"
)

type HatenaFetcher struct {
	client *http.Client
}

func NewHatenaFetcher(client *http.Client) *HatenaFetcher {
	if client == nil {
		client = &http.Client{Timeout: 20 * time.Second}
	}
	return &HatenaFetcher{client: client}
}

func (f *HatenaFetcher) Name() string { return "hatena" }

func (f *HatenaFetcher) FetchItems(ctx context.Context, _ string) ([]domain.SourceItem, error) {
	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, "https://b.hatena.ne.jp/hotentry/it.rss", nil)
	res, err := f.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		b, _ := io.ReadAll(res.Body)
		return nil, errors.New("hatena http error: " + string(b))
	}
	xmlBytes, _ := io.ReadAll(res.Body)
	xml := string(xmlBytes)
	items := reItemHatena.FindAllString(xml, -1)
	out := make([]domain.SourceItem, 0, len(items))
	for i, it := range items {
		title := decodeXML(matchTag(it, "title"))
		link := decodeXML(matchTag(it, "link"))
		desc := sanitizeText(decodeXML(matchTag(it, "description")))
		date := matchTag(it, "dc:date")
		if date == "" {
			date = time.Now().UTC().Format(time.RFC3339)
		}
		bc, _ := strconv.ParseFloat(matchTag(it, "hatena:bookmarkcount"), 64)
		if title == "" || link == "" {
			continue
		}
		out = append(out, domain.SourceItem{
			Source:         domain.SourceHatena,
			ID:             strconv.Itoa(i + 1),
			Title:          title,
			URL:            link,
			Score:          bc,
			Comments:       0,
			PublishedAt:    normalizeRFC3339(date),
			ContentSnippet: desc,
		})
		if len(out) >= 60 {
			break
		}
	}
	return out, nil
}

var reItemHatena = regexp.MustCompile(`(?s)<item\b.*?</item>`)

func matchTag(xml, tag string) string {
	re := regexp.MustCompile(`(?is)<` + regexp.QuoteMeta(tag) + `[^>]*>(.*?)</` + regexp.QuoteMeta(tag) + `>`)
	m := re.FindStringSubmatch(xml)
	if len(m) < 2 {
		return ""
	}
	return strings.TrimSpace(stripCDATA(m[1]))
}

func stripCDATA(s string) string {
	s = strings.TrimSpace(s)
	s = strings.TrimPrefix(s, "<![CDATA[")
	s = strings.TrimSuffix(s, "]]>")
	return s
}

func decodeXML(s string) string {
	r := strings.NewReplacer("&amp;", "&", "&lt;", "<", "&gt;", ">", "&quot;", "\"", "&#39;", "'")
	s = r.Replace(s)
	s = decodeNumericEntities(s)
	return s
}

func decodeNumericEntities(s string) string {
	reHex := regexp.MustCompile(`&#x([0-9a-fA-F]+);`)
	s = reHex.ReplaceAllStringFunc(s, func(m string) string {
		h := reHex.FindStringSubmatch(m)[1]
		n, err := strconv.ParseInt(h, 16, 32)
		if err != nil {
			return m
		}
		return string(rune(n))
	})
	reDec := regexp.MustCompile(`&#([0-9]+);`)
	s = reDec.ReplaceAllStringFunc(s, func(m string) string {
		d := reDec.FindStringSubmatch(m)[1]
		n, err := strconv.ParseInt(d, 10, 32)
		if err != nil {
			return m
		}
		return string(rune(n))
	})
	return s
}

func sanitizeText(s string) string {
	re := regexp.MustCompile(`<[^>]+>`)
	s = re.ReplaceAllString(s, " ")
	s = strings.Join(strings.Fields(s), " ")
	if len(s) > 600 {
		return s[:600]
	}
	return s
}

func normalizeRFC3339(v string) string {
	if t, err := time.Parse(time.RFC3339, v); err == nil {
		return t.UTC().Format(time.RFC3339)
	}
	if t, err := time.Parse(time.RFC1123Z, v); err == nil {
		return t.UTC().Format(time.RFC3339)
	}
	if t, err := time.Parse(time.RFC1123, v); err == nil {
		return t.UTC().Format(time.RFC3339)
	}
	return time.Now().UTC().Format(time.RFC3339)
}

File: src/adapters/sources/hatenaFetcher.ts
import { SourceItem } from "../../domain/model.js";
import { SourceFetcher } from "../../ports/sourceFetcher.js";
import { withRetry } from "../../shared/retry.js";

export class HatenaFetcher implements SourceFetcher {
  readonly name = "hatena";

  async fetchItems(): Promise<SourceItem[]> {
    const url = "https://b.hatena.ne.jp/hotentry/it.rss";
    const xml = await withRetry(
      async () => {
        const res = await fetch(url);
        if (!res.ok) {
          throw new Error(`HTTP ${res.status}: ${await res.text()}`);
        }
        return res.text();
      },
      2,
      500
    );
    const entries = parseHatenaRss(xml);

    return entries.slice(0, 60).map((x, i) => ({
      source: "hatena",
      id: String(i + 1),
      title: x.title,
      url: x.link,
      score: x.bookmarks,
      comments: 0,
      publishedAt: x.date,
      contentSnippet: x.description
    }));
  }
}

function parseHatenaRss(
  xml: string
): Array<{ title: string; link: string; bookmarks: number; date: string; description: string }> {
  const items = xml.match(/<item\b[\s\S]*?<\/item>/g) ?? [];
  return items
    .map((item) => {
      const title = decodeXml(matchTag(item, "title") ?? "");
      const link = decodeXml(matchTag(item, "link") ?? "");
      const description = sanitizeText(decodeXml(matchTag(item, "description") ?? ""));
      const date = matchTag(item, "dc:date") ?? new Date().toISOString();
      const bookmarkRaw = matchTag(item, "hatena:bookmarkcount") ?? "0";
      const bookmarks = Number(bookmarkRaw) || 0;
      return { title, link, date, bookmarks, description };
    })
    .filter((x) => x.title.length > 0 && x.link.length > 0);
}

function matchTag(xml: string, tag: string): string | undefined {
  const m = xml.match(new RegExp(`<${tag}>([\\s\\S]*?)<\\/${tag}>`, "i"));
  return m?.[1]?.trim();
}

function decodeXml(text: string): string {
  return text
    .replace(/&#x([0-9a-fA-F]+);/g, (_, hex: string) => String.fromCodePoint(parseInt(hex, 16)))
    .replace(/&#([0-9]+);/g, (_, num: string) => String.fromCodePoint(parseInt(num, 10)))
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, "\"")
    .replace(/&#39;/g, "'");
}

function sanitizeText(text: string): string {
  return text.replace(/<[^>]+>/g, " ").replace(/\s+/g, " ").trim().slice(0, 600);
}
